### PR TITLE
fix: Circular linked list

### PR DESCRIPTION
**Pull request by polarvoid**
#### Description of Change
Rewrote Circular Linked List, added documentation and tests

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: I rewrote the program, added doc and tests

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1825"></a>

### CODE REVIEW

File: operations_on_datastructures/circular_linked_list.cpp [hunk 1]
```diff
@@ -1,97 +1,275 @@
...
+    }
+};
+
+}  // namespace circular_linked_list
+
+}  // namespace operations_on_datastructures
+
+/**
+ * @namespace tests
+ * @brief Testcases to check Circular Linkedd List.
```
Comment (Panquesito7) [inline operations_on_datastructures/circular_linked_list.cpp:170]:
```suggestion
 * @brief Testcases to check Circular Linked List.
```

File: operations_on_datastructures/circular_linked_list.cpp [hunk 2]
```diff
@@ -1,97 +1,275 @@
...
+ * @namespace operations_on_datastructures
+ * @brief Operations on Data Structures
+ */
+namespace operations_on_datastructures {
 
-void insert(int x) {
-    node *t = start;
+/**
+ * @namespace circular_linked_list
+ * @brief Functions for the Circular Linked List implementation
```
Comment (Panquesito7) [inline operations_on_datastructures/circular_linked_list.cpp:24]:
```suggestion
 * @brief Functions for the [Circular Linked List](https://www.geeksforgeeks.org/circular-linked-list/) implementation
```

File: operations_on_datastructures/circular_linked_list.cpp [hunk 3]
```diff
@@ -1,97 +1,275 @@
...
-        start = n;
-        n->next = start;
+/**
+ * @brief A Node struct that represents a single Node in a Binary Tree
+ */
+struct Node {
+    int64_t data;  ///< The value of the Node
+    Node* next;    ///< The Node's successor
+    /**
+     * @brief Creates a new Node with some initial data
```
Comment (Panquesito7) [inline operations_on_datastructures/circular_linked_list.cpp:35]: Missing parameter documentation (`_data`).

File: operations_on_datastructures/circular_linked_list.cpp [hunk 4]
```diff
@@ -1,97 +1,275 @@
...
+/**
+ * @brief Function to test the correctness of the Circular Linked List
+ * @returns void
+ */
+static void test() {
+    tests::test1();
+    tests::test2();
+    tests::test3();
+    tests::test4();
+    tests::test5();
```
Comment (Panquesito7) [inline operations_on_datastructures/circular_linked_list.cpp:265]:
```suggestion
    tests::test5();

    std::cout << "All tests have successfully passed!\n";
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Woah, amazing work! 👀💥

File: operations_on_datastructures/circular_linked_list.cpp (hunk 3 above)
Comment (polarvoid) [reply operations_on_datastructures/circular_linked_list.cpp:36]: Oooh, I missed that, thanks :)

**Review by Panquesito7 — DISMISSED** [review]
Amazing work! 👀 Thank you so much for your continued contributions! 😄👍🎉

**Review by amino19 — DISMISSED** [review]

**Review by ayaankhan98 — CHANGES_REQUESTED** [review]
Thanks for your contribution. Please fix the issue below.
```
=================================================================
==12380==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 16 byte(s) in 1 object(s) allocated from:
    #0 0x7f07cff18ca1 in operator new(unsigned long) /build/gcc/src/gcc/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x55f8c3a9a4d0 in operations_on_datastructures::circular_linked_list::CircularLinkedList::insert(long) /home/ayaan/test.cpp:90
    #2 0x55f8c3a98525 in tests::test1() /home/ayaan/test.cpp:189
    #3 0x55f8c3a99e6a in test /home/ayaan/test.cpp:265
    #4 0x55f8c3a99e8a in main /home/ayaan/test.cpp:277
    #5 0x7f07cefd4b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)

Indirect leak of 96 byte(s) in 6 object(s) allocated from:
    #0 0x7f07cff18ca1 in operator new(unsigned long) /build/gcc/src/gcc/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x55f8c3a9a4d0 in operations_on_datastructures::circular_linked_list::CircularLinkedList::insert(long) /home/ayaan/test.cpp:90
    #2 0x55f8c3a9a41b in operations_on_datastructures::circular_linked_list::CircularLinkedList::insert(std::vector<long, std::allocator<long> > const&) /home/ayaan/test.cpp:79
    #3 0x55f8c3a991c7 in tests::test3() /home/ayaan/test.cpp:221
    #4 0x55f8c3a99e74 in test /home/ayaan/test.cpp:267
    #5 0x55f8c3a99e8a in main /home/ayaan/test.cpp:277
    #6 0x7f07cefd4b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)

Indirect leak of 16 byte(s) in 1 object(s) allocated from:
    #0 0x7f07cff18ca1 in operator new(unsigned long) /build/gcc/src/gcc/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x55f8c3a9a4d0 in operations_on_datastructures::circular_linked_list::CircularLinkedList::insert(long) /home/ayaan/test.cpp:90
    #2 0x55f8c3a98a08 in tests::test2() /home/ayaan/test.cpp:205
    #3 0x55f8c3a99e6f in test /home/ayaan/test.cpp:266
    #4 0x55f8c3a99e8a in main /home/ayaan/test.cpp:277
    #5 0x7f07cefd4b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)

Indirect leak of 16 byte(s) in 1 object(s) allocated from:
    #0 0x7f07cff18ca1 in operator new(unsigned long) /build/gcc/src/gcc/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x55f8c3a9a4d0 in operations_on_datastructures::circular_linked_list::CircularLinkedList::insert(long) /home/ayaan/test.cpp:90
    #2 0x55f8c3a98a1c in tests::test2() /home/ayaan/test.cpp:206
    #3 0x55f8c3a99e6f in test /home/ayaan/test.cpp:266
    #4 0x55f8c3a99e8a in main /home/ayaan/test.cpp:277
    #5 0x7f07cefd4b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)

Indirect leak of 16 byte(s) in 1 object(s) allocated from:
    #0 0x7f07cff18ca1 in operator new(unsigned long) /build/gcc/src/gcc/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x55f8c3a9a4d0 in operations_on_datastructures::circular_linked_list::CircularLinkedList::insert(long) /home/ayaan/test.cpp:90
    #2 0x55f8c3a9979c in tests::test4() /home/ayaan/test.cpp:236
    #3 0x55f8c3a99e79 in test /home/ayaan/test.cpp:268
    #4 0x55f8c3a99e8a in main /home/ayaan/test.cpp:277
    #5 0x7f07cefd4b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)

Indirect leak of 16 byte(s) in 1 object(s) allocated from:
    #0 0x7f07cff18ca1 in operator new(unsigned long) /build/gcc/src/gcc/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x55f8c3a997a6 in tests::test4() /home/ayaan/test.cpp:237
    #2 0x55f8c3a99e79 in test /home/ayaan/test.cpp:268
    #3 0x55f8c3a99e8a in main /home/ayaan/test.cpp:277
    #4 0x7f07cefd4b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)

Indirect leak of 16 byte(s) in 1 object(s) allocated from:
    #0 0x7f07cff18ca1 in operator new(unsigned long) /build/gcc/src/gcc/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x55f8c3a9a4d0 in operations_on_datastructures::circular_linked_list::CircularLinkedList::insert(long) /home/ayaan/test.cpp:90
    #2 0x55f8c3a989f4 in tests::test2() /home/ayaan/test.cpp:204
    #3 0x55f8c3a99e6f in test /home/ayaan/test.cpp:266
    #4 0x55f8c3a99e8a in main /home/ayaan/test.cpp:277
    #5 0x7f07cefd4b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)

SUMMARY: AddressSanitizer: 192 byte(s) leaked in 12 allocation(s).
```

**Comment by polarvoid** [comment]
Hi, could you help me resolve this issue? I'm guessing this has something to do with freeing the memory of the Nodes once the list gets destroyed?

**Comment by polarvoid** [comment]
Okay so I think I got that memory leak fixed. I'm failing the code formatting tests now however, and it says to use smart pointers.

**Review by ayaankhan98 — APPROVED** [review]
LGTM!

**Comment by Panquesito7** [comment]
The [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/runs/4068571385?check_suite_focus=true) checks are now failing. What should we do? @ayaankhan98, @mishraabhinn, @amino19.

**Comment by ayaankhan98** [comment]
> The [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/runs/4068571385?check_suite_focus=true) checks are now failing. What should we do? @ayaankhan98, @mishraabhinn, @amino19.

just fix the errors ;)

**Comment by Panquesito7** [comment]
> > The [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/runs/4068571385?check_suite_focus=true) checks are now failing. What should we do? @ayaankhan98, @mishraabhinn, @amino19.
> 
> just fix the errors ;)

https://github.com/TheAlgorithms/C-Plus-Plus/pull/1825#issuecomment-955750603

**Comment by polarvoid** [comment]
### IT WORKS!
So I managed to pass all tests. I added copy and move constructors and assignment operators. I am a little unsure about the implementation though, since I did not know move constructors even existed so do check that part out 😄

**Review by ayaankhan98 — COMMENTED** [review]
```
test.cpp: In member function ‘operations_on_datastructures::circular_linked_list::CircularLinkedList& operations_on_datastructures::circular_linked_list::CircularLinkedList::operator=(const operations_on_datastr
uctures::circular_linked_list::CircularLinkedList&)’:
test.cpp:105:5: error: no return statement in function returning non-void [-Werror=return-type]
  104 |         }
  +++ |+    return *this;
  105 |     }
      |     ^
test.cpp: In member function ‘operations_on_datastructures::circular_linked_list::CircularLinkedList& operations_on_datastructures::circular_linked_list::CircularLinkedList::operator=(operations_on_datastructure
s::circular_linked_list::CircularLinkedList&&)’:
test.cpp:116:5: error: no return statement in function returning non-void [-Werror=return-type]
  115 |         other.end = nullptr;
  +++ |+                        return *this;
  116 |     }
      |     ^
cc1plus: all warnings being treated as errors

shell returned 1
```

**Review by ayaankhan98 — APPROVED** [review]
Nice :+1:

**Review by amino19 — APPROVED** [review]

**Review by Panquesito7 — APPROVED** [review]

**Pull request merged**